File: solutions/307/main.py
from typing import * 

class TreeNode(object):
    def __init__(self, value = 0):
        self.value = value 
        self.left = None 
        self.right = None 

class NumArray:

    def __init__(self, nums: List[int]):
        self.nums = nums 
        self.n = len(nums)
        if self.n == 0:
            return 
        
        self.seg_tree = self.construct_seg_tree(0, self.n - 1)

    def update(self, i: int, val: int):
        self.update_seg_tree(self.seg_tree, 0, self.n - 1, i, val)
        self.nums[i] = val 
        

    def sumRange(self, i: int, j: int):
        return self.tree_sum(self.seg_tree, 0, self.n-1, i, j)
    
    def construct_seg_tree(self, st, ed):
        tmp = TreeNode()
        mid = st + (ed - st) // 2 
        if st == ed:
            tmp.value = self.nums[st]
        else:
            tmp.left = self.construct_seg_tree(st, mid)
            tmp.right = self.construct_seg_tree(mid+1, ed)
            tmp.value = tmp.right.value + tmp.left.value 
        return tmp 

    def tree_sum(self, seg_tree, st, ed, i, j):
        if st >= i and ed <= j:
            return seg_tree.value 
        elif ed < i or st > j:
            return 0 
        else:
            mid = st + (ed - st) // 2 
        return self.tree_sum(seg_tree.left, st, mid, i, j) + \
            self.tree_sum(seg_tree.right, mid+1, ed, i, j)
        
    def update_seg_tree(self, seg_tree, st, ed, i, val):
        mid = st + (ed - st) // 2 
        diff = val - self.nums[i]
        if st == ed:
            seg_tree.value += diff 
        else:
            if i <= mid:
                seg_tree.value += diff 
                self.update_seg_tree(seg_tree.left, st, mid, i, val) 
            else:
                seg_tree.value += diff 
                self.update_seg_tree(seg_tree.right, mid+1, ed, i, val)
                
                
# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(i,val)
# param_2 = obj.sumRange(i,j)

File: solutions/307/main.cpp
#include <vector> 

class NumArray {
public:
    NumArray(vector<int>& nums) {
        n = nums.size();
        if (n == 0) {return;}
        this->nums = nums;
        segTree = constructSegTree(nums, 0, n-1);
    }
    
    void update(int i, int val) {
        updateSegTree(segTree, 0, n-1, i, val);
        nums[i] = val;
    }
    
    int sumRange(int i, int j) {
        return treeSum(segTree, 0, n-1, i, j);
    }
    
private:
    struct TreeNode
    {
        int value;
        TreeNode *left, *right;
    };
    TreeNode *segTree;
    int n;
    vector<int> nums;
    
    TreeNode *constructSegTree(const vector<int> &nums, int st, int ed){
        TreeNode *tnode = new TreeNode();
        if (st == ed){
            tnode->value = nums[st];
        }
        else{
            int mid = st + 1/2 * (ed - st);
            tnode->left = constructSegTree(nums, st, mid);
            tnode->right = constructSegTree(nums, mid+1, ed);
            tnode->value = tnode->left->value + tnode->right->value;
        }
        return tnode;
    }
    
    int treeSum(TreeNode *segTree, int st, int ed, int l, int r)
    {
        if (st >= l && ed <= r)
        {
            return segTree->value;
        }
        else if (ed < l || st > r)
        {
            return 0;
        }
        else {
            int mid = st + 1/2 * (ed - st);
            return treeSum(segTree->left, st, mid, l, r) + 
                treeSum(segTree->right, mid+1, ed, l, r);
        }
    }
    
    void updateSegTree(TreeNode *segTree, int st, int ed, int i, int val){
        int mid = st + 1/2 * (ed - st);
        int diff = val - nums[i];
        if (st == ed){
            segTree->value += diff;
        }
        else if (i <= mid){
            segTree->value += diff;
            updateSegTree(segTree->left, st, mid, i, val);
        }
        else {
            segTree->value += diff;
            updateSegTree(segTree->right, mid+1, ed, i, val);
        }
    }    
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * obj->update(i,val);
 * int param_2 = obj->sumRange(i,j);
 */